api/interceptor/observer_http.go

```diff
@@ -0,0 +1,56 @@
+package interceptor
+
+import (
+	"io"
+	"net/http"
+)
+
+type bodyReader struct {
+	base      io.ReadCloser
+	bytesRead *int64
+}
+
+func (r bodyReader) Read(p []byte) (int, error) {
+	n, err := r.base.Read(p)
+	*r.bytesRead += int64(n)
+	return n, err
+}
+
+func (r bodyReader) Close() error {
+	return r.base.Close()
+}
+
+type responseWriteFlusher interface {
+	http.ResponseWriter
+	http.Flusher
+}
+
+type responseController struct {
+	base responseWriteFlusher
+	rm   *RequestMetadata
+}
+
+var _ responseWriteFlusher = (*responseController)(nil)
+
+func (r responseController) Header() http.Header {
+	return r.base.Header()
+}
+
+func (r responseController) Write(p []byte) (int, error) {
+	n, err := r.base.Write(p)
+	r.rm.bytesSent += int64(n)
+	return n, err
+}
+
+func (r responseController) WriteHeader(statusCode int) {
```

Comment:
what if someone does not call WriteHeader? or is it guaranteed that WriteHeader is called , even if the "user" (some http handler implementation) does not call it?

Reply:
This will always be called. While you can call `Write()` directly this will call `WriteHeader(http.StatusOK)` in case it has not yet been called.